devise を使ってユーザー認証を実装する

takao0206 opened the pull request:
ユーザー認証機能を実装しました。

takao0206 committed:
ユーザー認証機能を実装した。

補足
- ログインが必要ないホーム画面を作成して、最初のアクセス時に表示されるようにした。

takao0206 committed:
`users_controller`の中で日本語で書かれた文章を`ja.yml`に移動。

yoshitsugu commented on config/locales/devise_ja.yml:
```diff
@@ -0,0 +1,12 @@
+ja:
+  devise:
```
`devise-i18n` gemを使っていれば勝手に
https://github.com/tigrish/devise-i18n/blob/master/rails/locales/ja.yml
こちらの翻訳ファイルを使ってくれるのですが、カスタイマイズしたい、などの特別な事情があったでしょうか？

takao0206 replied:
`devise-i18n/rails/locales/ja.yml`の確認不足でした。申し訳ございません。
`devise-i18n/rails/locales/ja.yml`を使うようにして、`views/devise/registrations/`下のファイルを修正したいと思います。
（`devise_ja.yml`は不要になると思いますので、削除したいと思います）
ご指摘いただき、ありがとうございました。

yoshitsugu commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+    </tbody>
+  </table>
+</div>
+<%= link_to t('views.common.edit', name: User.model_name.human), edit_user_path, class: "navbar-link" %>
```
編集不可のユーザーに対してはリンクを表示しなくて大丈夫です。
もしくはグローバルナビゲーションに「プロフィールを編集」のような、そもそも自分自身を対象とした編集しかしないようなリンクを作りましょう。

takao0206 replied:
アドバイスいただき、ありがとうございました。
グローバルナビゲーションに自分自身をできるリンクを表示するように変更したいと思います。

yoshitsugu commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,27 @@
+      </tr>
+      <tr>
+        <th><%= User.human_attribute_name(:created_at) %></th>
+        <td><%= @user.created_at.strftime('%Y年%m月%d日') %></td>
```
`strftime` はRailsではあまり使いません。 `I18n.l` を使いましょう。
参考: https://qiita.com/jnchito/items/831654253fb8a958ec25

takao0206 replied:
`I18n.l`のご紹介、ありがとうございました。
今後は、`l`や`I18n.l`を使いたいと思います。

yoshitsugu commented on app/views/users/_form.html.erb:
自分自身しか編集できない以上、このviewは不要そうに思いますが、どうでしょうか

takao0206 replied:
おっしゃる通り、`app/views/users/_form.html.erb`は不要だと思います。
削除したいと思います。

yoshitsugu commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,26 @@
+  def show; end
+
+  def edit
+    if current_user == @user
+      redirect_to edit_user_registration_path
```
あまりリダイレクトする意味もないので、そもそも `edit_user_registrration_path` のみ使うようにしましょう。

takao0206 replied:
`edit_user_registration_path`のみを使う方法を調べました。
controllerではなくviewsに直接`edit_user_registration_path`を記載する方がシンプルと思いましたので、
一度その方法で修正してみたいと思います。

yoshitsugu reviewed (commented):
コメントしました！

takao0206 committed:
ご指摘事項の修正（4点）

- deviseの日本語訳について、`devise-i18n/rails/locales/ja.yml`を使うように変更（不要となった`devise_ja.yml`は削除）。
- 日本語で日付を表示する方法として、`I18n.l`を使用するように変更。
- `app/views/users/_form.html.erb`は不要なため、削除。
- ユーザーを編集するリンク`edit_user_registration_path`を`views/layouts/application.html.erb`に追記（不要となった`users_controller`のメソッド`edit`を削除）。

takao0206 commented:
アドバイスいただきました点、修正致しました。
ご都合の良いときにご確認いただければ幸いです。

yoshitsugu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+
+  # ==> Configuration for :validatable
+  # Range for password length.
+  config.password_length = 6..128
```
今回はプラクティスなのでよいですが、本来パスワードの長さは最低10文字以上にするなどもう少し長いもののほうがセキュアです。

yoshitsugu reviewed (approved):
OKです！一点コメントしましたので参考までに。

takao0206 commented:
パスワードの長さのアドバイス、ありがとうございます。
今後、パスワードの長さは10桁以上を目安にしたいと思います。
